perses/rjmc/topology_proposal.py

```diff
@@ -1563,6 +1863,105 @@ def compute_state_key(self, topology):
+            else:
+                pass
+
+        #loop over new topology
+        for bond in topology_proposal.new_topology.bonds():
+            if bond[0].index in topology_proposal._new_alchemical_atoms and bond[1].index in topology_proposal.new_environment_atoms:
+                new_alchemical_to_nonalchemical_bonds.append((bond[0], bond[1]))
+            elif bond[1].index in topology_proposal._new_alchemical_atoms and bond[0].index in topology_proposal.new_environment_atoms:
+                new_alchemical_to_nonalchemical_bonds.append((bond[1], bond[0]))
+            else:
+                pass
+
+        assert len(old_alchemical_to_nonalchemical_bonds) == len(new_alchemical_to_nonalchemical_bonds), f"the number of alchemical to nonalchemical bonds in old and new topologies is not equal"
+
+        #assert that all of the alchemical atoms (bonded to nonalchemical atoms) are 'core'
+        new_to_old_pair = {}
+        for alch_atom, nonalch_atom in old_alchemical_to_nonalchemical_bonds:
+            assert alch_atom.index in list(topology_proposal.core_new_to_old_atom_map.values()), f"the old alchemical atom ({alch_atom.index}) is not a core atom!"
+            assert nonalch_atom.index in topology_proposal.old_environment_atoms, f"the nonalchemical atom ({nonalch_atom.index}) is not an environment atom!"
+            appropriate_new_bonds = [bond for bond in new_alchemical_to_nonalchemical_bonds if bond[0].index == topology_proposal.old_to_new_atom_map[alch_atom.index] and bond[1].index == topology_proposal.old_to_new_atom_map[nonalch_atom.index]]
+            assert len(appropriate_new_bonds) == 1, f"there is no match between the old bond to the new bond"
+            new_to_old_pair[appropriate_new_bonds[0]] = (alch_atom, nonalch_atom)
+
+
+        #if there is at least one alchemical to nonalchemical bond, then we should check the system...
+        old_forces = {force.__class__.__name__ : force for force in [old_system.getForce(index) for index in range(old_system.getNumForces())]}
+        new_forces = {force.__class__.__name__ : force for force in [new_system.getForce(index) for index in range(new_system.getNumForces())]}
+        assert set(list(old_forces.keys())) == set(list(new_forces.keys())), f"the old and new forces do not match: {old_forces, new_forces}"
+
+        for new_bond, old_bond in new_to_old_pair.items():
+            new_atom_idx1, new_atom_idx2 = new_bond[0].index, new_bond[1].index
+            old_atom_idx1, old_atom_idx2 = old_bond[0].index, old_bond[1].index
+            _logger.debug(f"\titerating through new atoms {new_atom_idx1, new_atom_idx2} with old atoms {old_atom_idx1, old_atom_idx2}")
+
+            if 'HarmonicBondForce' in list(old_forces.keys()):
+                old_bond_terms = [param for param in [old_forces['HarmonicBondForce'].getBondParameters(bond_index) for bond_index in range(old_forces['HarmonicBondForce'].getNumBonds())] if set((old_atom_idx1, old_atom_idx2)) == set(param[:2])]
```

Comment:
Can we have a one line function called `_get_terms()` or something? Might make this block more readable